// tdv-core/dist/src/model/metadata/PropertyMetadata.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _PropertyMetadata_runtimeValue;
Object.defineProperty(exports, "__esModule", { value: true });
const MetadataProcessor_1 = __importDefault(require("../processor/MetadataProcessor"));
class PropertyMetadata {
    get parent() {
        return this._owner;
    }
    get name() {
        return this._name;
    }
    get clazz() {
        return this._clazz;
    }
    get type() {
        return this._type;
    }
    constructor(clazz, field, runtimeValue) {
        _PropertyMetadata_runtimeValue.set(this, void 0);
        __classPrivateFieldSet(this, _PropertyMetadata_runtimeValue, runtimeValue, "f");
        this._owner = clazz;
        this._name = field;
        this._clazz = this.buildClass();
        this._type = this.buildTypeGroup();
    }
    buildTypeGroup() {
        const p = MetadataProcessor_1.default.fromClass(this._owner);
        const vp = p.getValidationProcessor(this._name);
        if (!vp.isInitialTypeGroup) {
            return vp.typeGroup;
        }
        const value = __classPrivateFieldGet(this, _PropertyMetadata_runtimeValue, "f");
        return Array.isArray(value)
            ? vp.constructorCreator
                ? "OBJECT_ARRAY"
                : "PRIMITIVE_ARRAY"
            : vp.constructorCreator
                ? "OBJECT"
                : "PRIMITIVE";
    }
    buildClass() {
        var _a, _b, _c;
        return ((_c = (_b = (_a = MetadataProcessor_1.default.fromClass(this._owner)
            .getValidationProcessor(this._name)).constructorCreator) === null || _b === void 0 ? void 0 : _b.call(_a)) !== null && _c !== void 0 ? _c : null);
    }
}
_PropertyMetadata_runtimeValue = new WeakMap();
exports.default = PropertyMetadata;
