// tdv-core/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validators = exports.setLocale = exports.makeValidator = exports.getLocale = exports.Rule = exports.EntityProcessor = void 0;
const validator_facade_1 = require("./src/decorators/facade/validator.facade");
Object.defineProperty(exports, "makeValidator", { enumerable: true, get: function () { return validator_facade_1.makeValidator; } });
const Locale_1 = require("./src/messages/model/Locale");
Object.defineProperty(exports, "getLocale", { enumerable: true, get: function () { return Locale_1.getLocale; } });
Object.defineProperty(exports, "setLocale", { enumerable: true, get: function () { return Locale_1.setLocale; } });
const EntityProcessor_1 = __importDefault(require("./src/model/processor/EntityProcessor"));
exports.EntityProcessor = EntityProcessor_1.default;
const validators_1 = __importDefault(require("./validators"));
exports.validators = validators_1.default;
const Rule_1 = __importDefault(require("./validators/any/Rule"));
exports.Rule = Rule_1.default;
