// tdv-core/dist/validators/date/FutureDate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const validator_facade_1 = require("../../src/decorators/facade/validator.facade");
const ErrorMessage_1 = __importDefault(require("../../src/messages/impl/ErrorMessage"));
const decorator_utils_1 = require("../../src/utils/decorator.utils");
function isFutureDate(date) {
    // TODO: Maybe bump nullable validity to higher hierarchy
    return (0, decorator_utils_1.evaluateNullableValidity)(date, (d) => {
        const currentDate = new Date();
        return d.getTime() > currentDate.getTime();
    });
}
function FutureDate(props) {
    return (0, validator_facade_1.makeValidator)({
        groups: (0, decorator_utils_1.extractGroups)(props),
        isValid: (date) => ({
            key: "FutureDate",
            message: (0, decorator_utils_1.extractMessage)(props, ErrorMessage_1.default.FutureDate(date)),
            valid: isFutureDate(date),
        }),
    });
}
exports.default = FutureDate;
