// tdv-core/dist/validators/date/TodayDate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const validator_facade_1 = require("../../src/decorators/facade/validator.facade");
const ErrorMessage_1 = __importDefault(require("../../src/messages/impl/ErrorMessage"));
const decorator_utils_1 = require("../../src/utils/decorator.utils");
function isTodayDate(date) {
    return (0, decorator_utils_1.evaluateNullableValidity)(date, (d) => {
        const currentDate = new Date();
        return (d.getDate() === currentDate.getDate() &&
            d.getMonth() === currentDate.getMonth() &&
            d.getFullYear() === currentDate.getFullYear());
    });
}
function TodayDate(props) {
    return (0, validator_facade_1.makeValidator)({
        groups: (0, decorator_utils_1.extractGroups)(props),
        isValid: (date) => ({
            key: "TodayDate",
            message: (0, decorator_utils_1.extractMessage)(props, ErrorMessage_1.default.TodayDate(date)),
            valid: isTodayDate(date),
        }),
    });
}
exports.default = TodayDate;
