// tdv-core/dist/validators/array/ArrayUnique.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const validator_facade_1 = require("../../src/decorators/facade/validator.facade");
const ErrorMessage_1 = __importDefault(require("../../src/messages/impl/ErrorMessage"));
const object_utils_1 = require("../../src/utils/object.utils");
const array_utils_1 = require("../../src/utils/array.utils");
const decorator_utils_1 = require("../../src/utils/decorator.utils");
const DEFAULTS = {
    hash: object_utils_1.hash,
    groups: [],
    message: ErrorMessage_1.default.ArrayUnique(),
};
function ArrayUnique(props = DEFAULTS) {
    var _a;
    const hashFn = typeof props === "string" ? object_utils_1.hash : (_a = props.hash) !== null && _a !== void 0 ? _a : object_utils_1.hash;
    return (0, validator_facade_1.makeValidator)({
        groups: (0, decorator_utils_1.extractGroups)(props),
        isValid: (array) => ({
            key: "ArrayUnique",
            message: (0, decorator_utils_1.extractMessage)(props, ErrorMessage_1.default.ArrayUnique()),
            valid: (0, array_utils_1.isArrayUnique)(array !== null && array !== void 0 ? array : [], (obj1, obj2) => hashFn(obj1) === hashFn(obj2)),
        }),
    });
}
exports.default = ArrayUnique;
